signer/remote_signer.go

```diff
@@ -73,98 +75,82 @@ func (rs *ReconnRemoteSigner) OnStop() {
 }
 
 func (rs *ReconnRemoteSigner) establishConnection(ctx context.Context) (net.Conn, error) {
+	ctx, cancel := context.WithTimeout(ctx, connRetrySec*time.Second)
+	defer cancel()
+
 	proto, address := cometnet.ProtocolAndAddress(rs.address)
 	netConn, err := rs.dialer.DialContext(ctx, proto, address)
 	if err != nil {
-
 		return nil, fmt.Errorf("dial error: %w", err)
 	}
 
 	conn, err := cometp2pconn.MakeSecretConnection(netConn, rs.privKey)
 	if err != nil {
+		netConn.Close()
 		return nil, fmt.Errorf("secret connection error: %w", err)
 	}
 
 	return conn, nil
 }
 
-type conWrapper struct {
-	conn net.Conn
-}
-
-func (rs *ReconnRemoteSigner) attemptConnection(ctx context.Context, cw *conWrapper, connected chan<- bool) {
-	conn, err := rs.establishConnection(ctx)
-	if err != nil {
-		sentryConnectTries.Add(float64(1))
-		totalSentryConnectTries.Inc()
-
-		rs.Logger.Error("Error establishing connection", "err", err)
-		return
-	}
-
-	cw.conn = conn
-
-	sentryConnectTries.Set(0)
-
-	rs.Logger.Info("Connected to Sentry", "address", rs.address)
-	connected <- true
-
-}
-
 // main loop for ReconnRemoteSigner
 func (rs *ReconnRemoteSigner) loop(ctx context.Context) {
-	var cw conWrapper
+	var conn net.Conn
 	for {
 		if !rs.IsRunning() {
-			if cw.conn != nil {
-				if err := cw.conn.Close(); err != nil {
-					rs.Logger.Error("Close", "err", err.Error()+"closing listener failed")
-				}
-			}
+			rs.closeConn(conn)
 			return
 		}
 
-		ticker := time.NewTicker(1 * time.Second)
-		connected := make(chan bool)
-
-	ConnLoop:
-		for cw.conn == nil {
-			ctx, cancel := context.WithCancel(ctx)
-			defer cancel()
-			go rs.attemptConnection(ctx, &cw, connected)
-			select {
-			case <-connected:
-				break ConnLoop
-			case <-ticker.C:
-				cancel()
-				rs.Logger.Info("Retrying", "sleep (s)", 1, "address", rs.address)
+		for conn == nil {
+			var err error
+			conn, err = rs.establishConnection(ctx)
```

Comment:
I'm not sure if a backoff would help here. For a timeout-driven failure mode, I would usually expect to see an increasing timeout. On the other hand, two seconds is already gratuitous; if it fails in 2 seconds I don't think increasing it will help.

Reply:
tendermint has a 3 second window for connections when the privval interface is enabled. If a remote signer does not connect during that timeframe, it panics, so a backoff beyond 2 seconds wouldn't work well unfortunately.